### PR TITLE
fix voc07+12 training

### PR DESCRIPTION
**Pull request by Nioolek**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Fixed an issue where VOC07 +12 dataset could not be trained.
`ConcatDataset` in `mmengine`  judge whether datasets metainfo is equal.
`VOC07 DATASET_TYPE` is `VOC2007`, but `VOC12 DATASET_TYPE` is `VOC2012`.
580562bf8e6964e6e9435aedd1bf4f7
This resulted in an inability to train when using the VOC07+12 dataset:
image

## Modification

Methods:
- add `def metainfo(self)` in `VOCDataset` to delete `DATASET_TYPE`.
- Dont judge and warn in `voc_metric.py`

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/datasets/voc.py [hunk 1]
```diff
@@ -29,3 +31,17 @@ def __init__(self, **kwargs):
             self._metainfo['DATASET_TYPE'] = 'VOC2012'
```
Comment (BIGWangYuDong) [inline mmdet/datasets/voc.py:31]: `self._metainfo['DATASET_TYPE'] = 'VOC'

File: mmdet/datasets/voc.py [hunk 2]
```diff
@@ -29,3 +31,17 @@ def __init__(self, **kwargs):
...
         else:
             self._metainfo['DATASET_TYPE'] = None
+
+    @property
+    def metainfo(self) -> dict:
```
Comment (BIGWangYuDong) [inline mmdet/datasets/voc.py:36]: If we use self._metainfo['DATASET_TYPE']= 'VOC', this logic can be deleted

File: mmdet/evaluation/metrics/voc_metric.py [hunk 3]
```diff
@@ -121,19 +121,7 @@ def compute_metrics(self, results: list) -> dict:
...
-                    warnings.warn('Pascal VOC2007 uses `11points` as default '
-                                  'evaluate mode, but you are using '
-                                  f'{self.eval_mode}.')
-                elif dataset_type == 'VOC2012' and self.eval_mode != 'area':
-                    warnings.warn('Pascal VOC2012 uses `area` as default '
-                                  'evaluate mode, but you are using '
-                                  f'{self.eval_mode}.')
-            else:
-                dataset_name = self.dataset_meta['CLASSES']
+            dataset_name = self.dataset_meta['CLASSES']
```
Comment (BIGWangYuDong) [inline mmdet/evaluation/metrics/voc_metric.py:124]: We can assert dataset_type=='VOCl' and dataset_name = 'voc'

**Comment by BIGWangYuDong** [comment]
@Nioolek Thanks for your contribution, Please remember to sign the CLA first.
image

File: mmdet/evaluation/metrics/voc_metric.py [hunk 4]
```diff
@@ -121,19 +121,7 @@ def compute_metrics(self, results: list) -> dict:
         eval_results = OrderedDict()
         if self.metric == 'mAP':
             assert isinstance(self.iou_thrs, list)
-            dataset_type = self.dataset_meta.get('DATASET_TYPE')
-            if dataset_type in ['VOC2007', 'VOC2012']:
```
Comment (BIGWangYuDong) [inline mmdet/evaluation/metrics/voc_metric.py:125]: Add a todo list here, we need find a more elegant way, or even fix the potential bug in mmengine later

File: mmdet/datasets/voc.py [hunk 5]
```diff
@@ -1,4 +1,6 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+import copy
```
Comment (BIGWangYuDong) [inline mmdet/datasets/voc.py:2]: seems unnecessary

**Review by BIGWangYuDong — APPROVED** [review]
LGTM

**Comment by BIGWangYuDong** [comment]
Kindly ping @HAOCHENYE have a look, maybe we can add `ignore_key` in ConcatDataset to avoid this error

**Comment by ZwwWayne** [comment]
Hi @Nioolek ,

Thanks for your kind PR. It seems that your branch is out of date and causes much difference.

Now, we require all the PRs to be merged into the development branch, i.e., dev branch for MMDet 2.x and dev-3.x branch for MMDet 3.x. 

If your modification is not checkout from these two branches, they may cause conflicts and you should add your modification based on these two branches.

If your modification is checkout from these two branches, your branch might be out of date. In such a case
Please rebase your branch to the target dev or dev-3.x branch. You can simply do that by modifying your pull behavior:

```
git config --local --add pull.rebase true  # make rebase a default behavior in pull
```

Then you can git pull and rebase the dev branch by

```
git pull dev-3.x  # if your modification is for MMDet 3.x
git pull dev  # if your modification is for MMDet 2.x
```

After rebase, you might need to add `--force` option when pushing code, e.g.,

```
git push [remote name] [you branch] --force
```

**Comment by ZwwWayne** [comment]
Since the issue will be resolved in MMEngine, this PR is closed.

**Pull request closed without merging**